storage/scylladb_permanode.c

```diff
@@ -0,0 +1,614 @@
+ * All Rights Reserved.
+ * This is free software; you can redistribute it and/or modify it under the
+ * terms of the MIT license. A copy of the license can be found in the file
+ * "LICENSE" at the root of this distribution.
+ */
+
+#include "scylladb_permanode.h"
+#include "common/trinary/trit_long.h"
+#define logger_id scylladb_logger_id
+
+#define ADDRESS_OFFSET NUM_TRYTES_SIGNATURE
+#define VALUE_OFFSET (ADDRESS_OFFSET + NUM_TRYTES_ADDRESS)
+#define OBSOLETE_TAG_OFFSET (VALUE_OFFSET + NUM_TRYTES_VALUE)
+#define TIMESTAMP_OFFSET (OBSOLETE_TAG_OFFSET + NUM_TRYTES_OBSOLETE_TAG)
+#define CURRENT_INDEX_OFFSET (TIMESTAMP_OFFSET + NUM_TRYTES_TIMESTAMP)
+#define LAST_INDEX_OFFSET (CURRENT_INDEX_OFFSET + NUM_TRYTES_CURRENT_INDEX)
+#define BUNDLE_OFFSET (LAST_INDEX_OFFSET + NUM_TRYTES_LAST_INDEX)
+#define TRUNK_OFFSET (BUNDLE_OFFSET + NUM_TRYTES_BUNDLE)
+#define BRNACH_OFFSET (TRUNK_OFFSET + NUM_TRYTES_TRUNK)
+
+static void print_error(CassFuture* future) {
+  const char* message;
+  size_t message_length;
+  cass_future_error_message(future, &message, &message_length);
+  ta_log_error("Error: %.*s\n", (int)message_length, message);
+}
+
+static status_t create_transaction_table(CassSession* session, bool need_truncate) {
+  status_t ret = SC_OK;
+
+  if (execute_query(session,
+                    "CREATE TABLE IF NOT EXISTS transaction ("
+                    "hash blob PRIMARY KEY,"
+                    "address blob,"
+                    "bundle blob,"
+                    "tag blob,"
+                    "approvees set<blob>,"
+                    "trytes blob );") != CASS_OK) {
+    ta_log_error("Fail to create table : transaction\n");
+    ret = SC_STORAGE_CASSANDRA_QUREY_FAIL;
+    goto exit;
+  }
+
+  if (execute_query(session, "CREATE INDEX IF NOT EXISTS ON transaction(bundle);") != CASS_OK) {
+    ta_log_error("Fail to create transaction table index : bundle\n");
+    return SC_STORAGE_CASSANDRA_QUREY_FAIL;
+  }
+
+  if (execute_query(session, "CREATE INDEX IF NOT EXISTS ON transaction(address);") != CASS_OK) {
+    ta_log_error("Fail to create transaction table index : address\n");
+    return SC_STORAGE_CASSANDRA_QUREY_FAIL;
+  }
+
+  if (execute_query(session, "CREATE INDEX IF NOT EXISTS ON transaction(tag);") != CASS_OK) {
+    ta_log_error("Fail to create transaction table index : tag\n");
+    return SC_STORAGE_CASSANDRA_QUREY_FAIL;
+  }
+
+  if (need_truncate) {
+    if (db_truncate_table(session, "transaction") != SC_OK) {
+      ta_log_error("Fail to truncate table : transaction\n");
+      return SC_STORAGE_CASSANDRA_QUREY_FAIL;
+    }
+  }
+exit:
+  return ret;
+}
+
+static CassStatement* ret_insert_transaction_statement(const CassPrepared* prepared, const tryte_t* hash,
+                                                       const tryte_t* trytes) {
+  CassStatement* statement = NULL;
+  statement = cass_prepared_bind(prepared);
+  cass_statement_bind_bytes_by_name(statement, "hash", (cass_byte_t*)hash, NUM_TRYTES_HASH);
```

Comment:
Do we need to check out the error situation here?

Reply:
I will amend the error check here and similar function calls.